File: backend/app/services/auth_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from fastapi import HTTPException, status

from app.models.user import User
from app.schemas.auth import UserRegister, UserLogin, UserResponse, TokenResponse
from app.utils.security import hash_password, verify_password, create_access_token


class AuthService:

    @staticmethod
    async def register(data: UserRegister, db: AsyncSession) -> TokenResponse:
        result = await db.execute(select(User).where(User.email == data.email))
        if result.scalar_one_or_none():
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Email already registered")

        user = User(
            email=data.email,
            hashed_password=hash_password(data.password),
            full_name=data.full_name,
            role="user",
        )
        db.add(user)
        await db.commit()
        await db.refresh(user)

        token = create_access_token(user.id, user.role)
        return TokenResponse(access_token=token, user=UserResponse.model_validate(user))

    @staticmethod
    async def login(data: UserLogin, db: AsyncSession) -> TokenResponse:
        result = await db.execute(select(User).where(User.email == data.email))
        user = result.scalar_one_or_none()

        if not user or not verify_password(data.password, user.hashed_password):
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password")

        if not user.is_active:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Account is deactivated")

        token = create_access_token(user.id, user.role)
        return TokenResponse(access_token=token, user=UserResponse.model_validate(user))

File: backend/src/services/authService.js
'use strict';
const { PrismaClient } = require('@prisma/client');
const { hashPassword, verifyPassword, createAccessToken } = require('../utils/security');
const { httpError } = require('../utils/helpers');

const prisma = new PrismaClient();

function formatUser(user) {
  return {
    id: user.id,
    email: user.email,
    full_name: user.fullName,
    role: user.role,
    is_active: user.isActive,
    created_at: user.createdAt,
  };
}

async function register({ email, password, fullName }) {
  const existing = await prisma.user.findUnique({ where: { email } });
  if (existing) throw httpError('Email already registered', 409);

  const user = await prisma.user.create({
    data: {
      email,
      hashedPassword: hashPassword(password),
      fullName: fullName || null,
      role: 'user',
    },
  });

  return {
    access_token: createAccessToken(user.id, user.role),
    token_type: 'bearer',
    user: formatUser(user),
  };
}

async function login({ email, password }) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user || !verifyPassword(password, user.hashedPassword)) {
    throw httpError('Invalid email or password', 401);
  }
  if (!user.isActive) throw httpError('Account is deactivated', 403);

  return {
    access_token: createAccessToken(user.id, user.role),
    token_type: 'bearer',
    user: formatUser(user),
  };
}

module.exports = { register, login, formatUser };